### PR TITLE
HBG 4 Cards

### PR DESCRIPTION
**Pull request by Simisays**
The only problem is with Flames of Moradin. When targeting a token, it will be destroyed but Forge will then create a dublicate of one of the non-tokens.

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/water_weird.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Name:Water Weird
+ManaCost:3 U
+Types:Creature Elemental Weird
+PT:3/4
+T:Mode$ DamageDone | ValidSource$ Card.Self | ValidTarget$ Opponent | CombatDamage$ True | TriggerZones$ Battlefield | Execute$ TrigReveal | TriggerDescription$ Whenever CARDNAME deals combat damage to a player, put a +1/+1 counter on CARDNAME if the top card of your library is a nonland card. Otherwise, you may mill a card.
+SVar:TrigReveal:DB$ PeekAndReveal | PeekAmount$ 1 | NoPeek$ True | NoReveal$ True | RememberPeeked$ True | SubAbility$ DBCounter
+SVar:DBCounter:DB$ PutCounter | ConditionDefined$ Remembered | ConditionPresent$ Card.nonLand | ConditionCompare$ EQ1 | CounterType$ P1P1 | CounterNum$ 1 | Execute$ TrigMill
+SVar:TrigMill:DB$ Mill | NumCards$ 1 | Defined$ You | ConditionDefined$ Remembered | ConditionNotPresent$ Card | SubAbility$ DBCleanup
+SVar:DBCleanup:DB$ Cleanup | ClearRemembered$ True
+Oracle:Whenever Water Weird deals combat damage to a player, put a +1/+1 counter on Water Weird if the top card of your library is a nonland card. Otherwise, you may mill a card.
```
Comment (paulsnoops) [inline forge-gui/res/cardsfolder/upcoming/water_weird.txt:10]: This should be the rebalanced script:
image

Comment (paulsnoops) [reply forge-gui/res/cardsfolder/upcoming/water_weird.txt:10]: For this card yes as it's an Alchemy only card and there is only one version on Arena.

File: forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+SVar:DestroyTargets:DB$ DestroyAll | ValidCards$ Remembered | SubAbility$ DBEffect
+SVar:DBConjure:DB$ MakeCard | DefinedName$ Remembered.nonToken | Zone$ Hand | RememberMade$ True 
+SVar:DBEffect:DB$ Effect | StaticAbilities$ PerpetualAbility | RememberObjects$ Remembered | Name$ Flames of Moradin's Perpetual Effect | Duration$ Permanent
+SVar:PerpetualAbility:Mode$ Continuous | RememberObjects$ Targeted | AddKeyword$ Alternative Cost: R | Affected$ Card.IsRemembered | AddTrigger$ TrigSac | EffectZone$ Command | AffectedZone$ Battlefield | Description$ This creature perpetually gains "You may pay {R} rather than pay this spell's mana cost" and "at the beginning of your end step, sacrifice this artifact." 
+SVar:TrigSac:DB$ DelayedTrigger | RememberObjects$ Remembered | Mode$ Phase | Phase$ End of Turn | Execute$ TrigSacrifice | TriggerDescription$ Sacrifice this artifact at your end step.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:9]: Not a DelayedTrigger

File: forge-gui/res/cardsfolder/upcoming/mind_spike.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Name:Mind Spike
+ManaCost:B
+Types:Sorcery
+A:SP$ Reveal | ValidTgts$ Opponent | RevealAllValid$ Card.nonLand+nonCreature+TargetedPlayerCtrl | RememberRevealed$ True | SubAbility$ DBDiscard | StackDescription$ SpellDescription | SpellDescription$ Target opponent reveals each noncreature, nonland card in their hand. You choose a card revealed this way. That player discards that card. You lose 2 life. If they didn't reveal a card this way, you draw a card.
+SVar:DBDiscard:DB$ ChangeZone | DefinedPlayer$ Targeted | Chooser$ You | Origin$ Hand | Destination$ Graveyard | ChangeNum$ 1 | ChangeType$ Card.nonLand+nonCreature+TargetedPlayerCtrl | Mandatory$ True | AlreadyRevealed$ True | RememberChanged$ True | AILogic$ BestCard | SubAbility$ DBLoseLife | StackDescription$ None
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mind_spike.txt:5]: ChangeZone is not Discard

File: forge-gui/res/cardsfolder/upcoming/water_weird.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Name:Water Weird
+ManaCost:3 U
+Types:Creature Elemental Weird
+PT:3/4
+T:Mode$ DamageDone | ValidSource$ Card.Self | ValidTarget$ Opponent | CombatDamage$ True | TriggerZones$ Battlefield | Execute$ TrigReveal | TriggerDescription$ Whenever CARDNAME deals combat damage to a player, put a +1/+1 counter on CARDNAME if the top card of your library is a nonland card. Otherwise, you may mill a card.
+SVar:TrigReveal:DB$ PeekAndReveal | PeekAmount$ 1 | NoPeek$ True | NoReveal$ True | RememberPeeked$ True | SubAbility$ DBCounter
+SVar:DBCounter:DB$ PutCounter | CounterNum$ 1 | Defined$ Self | ConditionDefined$ Remembered | ConditionPresent$ Card.nonLand | ConditionCompare$ EQ1 | CounterType$ P1P1 | SubAbility$ DBMill
+SVar:DBMill:DB$ Mill | NumCards$ 1 | Defined$ Player | Optional$ True | ConditionDefined$ Remembered | ConditionPresent$ Card.Land | ConditionCompare$ EQ1 | SubAbility$ DBCleanup
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/water_weird.txt:8]: Why `Defined$ Player`?

File: forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Name:Yuan-Ti Scaleshield
+ManaCost:2 G
+Types:Instant
+A:SP$ PumpAll | Cost$ 2 G | ValidCards$ Permanent.YouCtrl | KW$ Hexproof & Indestructible | SubAbility$ DBSeek | SpellDescription$ Permanents you control gain hexproof and indestructible until end of turn. Seek a creature card if an opponent has cast a spell with mana value 3 or less this turn.
+SVar:DBSeek:DB$ ChangeZone | ConditionCheckSVar$ X | ConditionSVarCompare$ GE1 | Origin$ Library | Destination$ Hand | AtRandom$ True | NoShuffle$ True | Mandatory$ True | NoLooking$ True | NoReveal$ True | ChangeType$ Card.Creature | ChangeNum$ 1 
+SVar:X:CountOpponent$ThisTurnCast_Card.cmcLE3+OppCtrl
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt:6]: CountOpponent doesn't exist...

File: forge-gui/res/cardsfolder/upcoming/mind_spike.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+Name:Mind Spike
+ManaCost:B
+Types:Sorcery
+A:SP$ Reveal | ValidTgts$ Opponent | RevealAllValid$ Card.nonLand+nonCreature+TargetedPlayerCtrl | RememberRevealed$ True | SubAbility$ DBDiscard | StackDescription$ SpellDescription | SpellDescription$ Target opponent reveals each noncreature, nonland card in their hand. You choose a card revealed this way. That player discards that card. You lose 2 life. If they didn't reveal a card this way, you draw a card.
+SVar:DBDiscard:DB$ Discard | ValidTgts$ Opponent | Mode$ Defined | RevealNumber$ X | NumCards$ 1 | SubAbility$ DBLoseLife
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/mind_spike.txt:5]: There is only one target above already

File: forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+Name:Flames of Moradin
+ManaCost:2 R R
+Types:Sorcery
+A:SP$ Destroy | TargetMin$ 0 | TargetMax$ 3  | Cost$ 3 R | ValidTgts$ Artifact | SubAbility$ DBRepeatEach | SpellDescription$ Destroy up to three target artifacts. Conjure a duplicate of each nontoken artifact destroyed this way into your hand. The duplicates perpetually gain “You may pay {R} rather than pay this spell's mana cost" and "at the beginning of your end step, sacrifice this artifact."
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:4]: `TargetMax$ 3..| Cost$ 3 R | ` - think this cost is bad? Just remove?

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:4]: I think + `TgtPrompt$ Select up to three target artifacts`? or does Forge already parse this well?

File: forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+Name:Flames of Moradin
+ManaCost:2 R R
+Types:Sorcery
+A:SP$ Destroy | TargetMin$ 0 | TargetMax$ 3  | Cost$ 3 R | ValidTgts$ Artifact | SubAbility$ DBRepeatEach | SpellDescription$ Destroy up to three target artifacts. Conjure a duplicate of each nontoken artifact destroyed this way into your hand. The duplicates perpetually gain “You may pay {R} rather than pay this spell's mana cost" and "at the beginning of your end step, sacrifice this artifact."
+SVar:DBRepeatEach:DB$ RepeatEach | RepeatSubAbility$ DBConjure | RepeatCards$ Card.Artifact+IsRemembered
+SVar:DBConjure:DB$ MakeCard | ConditionDefined$ Remembered | ConditionPresent$ Card | DefinedName$ Remembered.nonToken | Zone$ Hand | SubAbility$ DBEffect
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:6]: I think `ConditionPresent$ Card.nonToken | DefinedName$ Remembered` would be better here

File: forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+ManaCost:2 R R
+Types:Sorcery
+A:SP$ Destroy | TargetMin$ 0 | TargetMax$ 3  | Cost$ 3 R | ValidTgts$ Artifact | SubAbility$ DBRepeatEach | SpellDescription$ Destroy up to three target artifacts. Conjure a duplicate of each nontoken artifact destroyed this way into your hand. The duplicates perpetually gain “You may pay {R} rather than pay this spell's mana cost" and "at the beginning of your end step, sacrifice this artifact."
+SVar:DBRepeatEach:DB$ RepeatEach | RepeatSubAbility$ DBConjure | RepeatCards$ Card.Artifact+IsRemembered
+SVar:DBConjure:DB$ MakeCard | ConditionDefined$ Remembered | ConditionPresent$ Card | DefinedName$ Remembered.nonToken | Zone$ Hand | SubAbility$ DBEffect
+SVar:DBEffect:DB$ Effect | StaticAbilities$ PerpetualAbility | Name$ Flames of Moradin's Perpetual Effect | Duration$ Permanent
+SVar:PerpetualAbility:Mode$ Continuous | AddKeyword$ Alternative Cost: R | Affected$ Card.IsRemembered+nonToken | AddTrigger$ TrigSac | EffectZone$ Command | AffectedZone$ Battlefield | Description$ This creature perpetually gains "You may pay {R} rather than pay this spell's mana cost" and "at the beginning of your end step, sacrifice this artifact."
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:8]: `AddKeyword$ Alternative Cost: R` > `AddKeyword$ Alternative Cost:R`

File: forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt (hunk 7 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/flames_of_moradin.txt:4]: `“You` > `"You`

File: forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+Name:Yuan-Ti Scaleshield
+ManaCost:2 G
+Types:Instant
+A:SP$ PumpAll | Cost$ 2 G | ValidCards$ Permanent.YouCtrl | KW$ Hexproof & Indestructible | SubAbility$ DBSeek | SpellDescription$ Permanents you control gain hexproof and indestructible until end of turn. Seek a creature card if an opponent has cast a spell with mana value 3 or less this turn.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt:4]: looks good - no need for Cost$ though

File: forge-gui/res/cardsfolder/upcoming/water_weird.txt [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+SVar:TrigReveal:DB$ PeekAndReveal | PeekAmount$ 1 | NoPeek$ True | NoReveal$ True | RememberPeeked$ True | SubAbility$ DBCounter
+SVar:DBCounter:DB$ PutCounter | CounterNum$ 1 | Defined$ Self | ConditionDefined$ Remembered | ConditionPresent$ Card.nonLand | ConditionCompare$ EQ1 | CounterType$ P1P1 | SubAbility$ DBMill
+SVar:DBMill:DB$ Mill | NumCards$ 1 | Optional$ True | ConditionDefined$ Remembered | ConditionPresent$ Card.Land | ConditionCompare$ EQ1 | SubAbility$ DBCleanup
+A:AB$ Pump | Cost$ 1 U | NumAtt$ +1 | NumDef$ -1 | SpellDescription$ CARDNAME gets +1/-1 until end of turn. 
+SVar:DBCleanup:DB$ Cleanup | ClearRemembered$ True
+Oracle:Whenever Water Weird deals combat damage to a player, put a +1/+1 counter on Water Weird if the top card of your library is a nonland card. Otherwise, you may mill a card..\n{1}{U}: Water Weird gets +1/-1 until end of turn.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/water_weird.txt:11]: `card..`

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/water_weird.txt:11]: DeckHas:Ability$Counters

File: forge-gui/res/cardsfolder/upcoming/water_weird.txt [hunk 12]
```diff
@@ -8,4 +8,5 @@ SVar:DBCounter:DB$ PutCounter | CounterNum$ 1 | Defined$ Self | ConditionDefined
 SVar:DBMill:DB$ Mill | NumCards$ 1 | Optional$ True | ConditionDefined$ Remembered | ConditionPresent$ Card.Land | ConditionCompare$ EQ1 | SubAbility$ DBCleanup
 A:AB$ Pump | Cost$ 1 U | NumAtt$ +1 | NumDef$ -1 | SpellDescription$ CARDNAME gets +1/-1 until end of turn. 
 SVar:DBCleanup:DB$ Cleanup | ClearRemembered$ True
-Oracle:Whenever Water Weird deals combat damage to a player, put a +1/+1 counter on Water Weird if the top card of your library is a nonland card. Otherwise, you may mill a card..\n{1}{U}: Water Weird gets +1/-1 until end of turn.
+DeckHas:Ability$Counters
+Oracle:Whenever Water Weird deals combat damage to a player, put a +1/+1 counter on Water Weird if the top card of your library is a nonland card. Otherwise, you may mill a card\n{1}{U}: Water Weird gets +1/-1 until end of turn.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/water_weird.txt:12]: `card\n` > `card.\n`

File: forge-gui/res/cardsfolder/upcoming/mind_spike.txt (hunk 6 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/mind_spike.txt:5]: You want `Defined$ Targeted` instead of `ValidTgts` in this line

**Comment by Simisays** [comment]
Update:Mind Spike is working, thanks for all the help !

Regarding Flames of Moradin, I haven't been able to make Forge not crash when you target a token. If anyone has any ideas, feel free to suggest them. 

Finally, I've added Arcane Archery, which has the last HBG card I've completed before switching to DMC. I've used a delaytrigger for the Boon effect.

**Comment by Simisays** [comment]
@tool4ever Removed Flames of Morradin. Will make a new PR when @Northmoc comes back since he's really knowledgeable about the alchemy stuff. He'll be able to make it work somehow. 

The other four cards should be good to merge

File: forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt [hunk 13]
```diff
@@ -0,0 +1,7 @@
+Name:Yuan-Ti Scaleshield
+ManaCost:2 G
+Types:Instant
+A:SP$ PumpAll | ValidCards$ Permanent.YouCtrl | KW$ Hexproof & Indestructible | SubAbility$ DBSeek | SpellDescription$ Permanents you control gain hexproof and indestructible until end of turn. Seek a creature card if an opponent has cast a spell with mana value 3 or less this turn.
+SVar:DBSeek:DB$ ChangeZone | ConditionCheckSVar$ X | ConditionSVarCompare$ GE1 | Origin$ Library | Destination$ Hand | AtRandom$ True | NoShuffle$ True | Mandatory$ True | NoLooking$ True | NoReveal$ True | ChangeType$ Card.Creature | ChangeNum$ 1 
+SVar:X:Count$ThisTurnCast_Card.cmcLE3+OppCtrl
+Oracle:Permanents you control gain hexproof and indestructible until end of turn./nSeek a creature card if an opponent has cast a spell with mana value 3 or less this turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/yuan_ti_scaleshield.txt:7]: Wrong linebreak

File: forge-gui/res/cardsfolder/upcoming/water_weird.txt [hunk 14]
```diff
@@ -0,0 +1,12 @@
+Name:Water Weird
+ManaCost:3 U
+Types:Creature Elemental Weird
+PT:3/4
+T:Mode$ DamageDone | ValidSource$ Card.Self | ValidTarget$ Opponent | CombatDamage$ True | TriggerZones$ Battlefield | Execute$ TrigReveal | TriggerDescription$ Whenever CARDNAME deals combat damage to a player, put a +1/+1 counter on CARDNAME if the top card of your library is a nonland card. Otherwise, you may mill a card.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/water_weird.txt:5]: ValidTarget$ Player

**Review by tool4ever — APPROVED** [review]

**Pull request merged**